demucs/api.py

```diff
@@ -0,0 +1,834 @@
+            Separator.load_audios_to_model(track)
+            failures = list(i.message.separate('"')[1] for i in w)
+        ```
+        """
+        if self._model is None:
+            raise RuntimeError("Please load model first!")
+        for track, wav in self.load_audios(*tracks):
+            if wav is None:
+                warnings.warn(f'"{track}" read failed and will not be loaded to model.')
+                continue
+            self._file.append(track)
+            self._wav.append(wav)
+
+    def load_audios_setup(self):
+        """
+        Load audios specied in the command line to the Separator that can be used in
+        `Separator.separate_loaded_audio`.
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        None
+
+        Notes
+        -----
+        When an error encountered, this function will only warn and continue loading other audios.
+        The audio failed to load will not be added into the Separator. To get a list of audios
+        failed to be loaded, you can use the following codes:
+        ```python
+        import warnings
+        with warnings.catch_warnings(record=True) as w:
+            Separator.load_audios_setup()
+            failures = list(i.message.separate('"')[1] for i in w)
+        ```
+        """
+        self.load_audios_to_model(*self._tracks)
+
+    def clear_filelist(self):
+        """
+        Remove all the loaded audios in the Separator.
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        None
+        """
+        self._wav = []
+        self._file = []
+
+    def add_track(self, filename: str, wav: th.FloatTensor):
+        """
+        Add a loaded track into the separator.
+
+        Parameters
+        ----------
+        filename: A string for you to remember what the each audio is. (You can call it \
+            "identifier")
+        wav: Waveform of the audio. Should have 2 dimensions, the first is each audio channel, \
+            while the second is the waveform of each channel. \
+            e.g. `tuple(wav.shape) == (2, 884000)` means the audio has 2 channels.
+
+        Returns
+        -------
+        None
+
+        Notes
+        -----
+        Use this function with cautiousness. This function does not provide data verifying.
+        """
+        self._file.append(filename)
+        self._wav.append(wav)
+
+    def _separate_track(
+        self,
+        wav,
+        model: Optional[AnyModel] = None,
+        segment: Optional[float] = 0.0,
+        shifts: Optional[int] = None,
+        split: Optional[bool] = None,
+        overlap: Optional[float] = None,
+        transition_power=1.0,
+        device=None,
+        num_workers=None,
+        pool=None,
+        callback: Optional[Callable[[dict], None]] = None,
+        callback_arg: Optional[dict] = None,
+        progress: bool = False,
+    ) -> th.Tensor:
```

Comment:
it seems like you copied the `apply_model` function, the logic is fairly complex and should never be duplicated. is there any reason for not directly calling the original `apply_model` function?

Reply:
As you can see, I'd like to add a `callback` argument which would be called before and after each chunk is separated - in fact, I need this to make a progress bar in my GUI
Rewriting this function in my own program made my code a mess ...